### ISSUE TITLE
throw exception/debug warning if an index mismatch occurs at `product(P::ProjMPO,wf)`

### ISSUE THREAD

**Issue by orialb** [ITensor/ITensors.jl#389]
Currently if `product(P::ProjMPO,wf)` is called and there is an index mismatch between `P` and `wf` (due to the[ user not being careful ](https://github.com/ITensor/ITensors.jl/issues/386#issuecomment-632639197)and calling orthogonalize! for example) it will fail silently and just return an `ITensor` with more indices which could be confusing to debug (of course the iterative method later on would fail).

It could be nice to throw an exception or at least some debug warning to indicate that something went wrong, e.g. something like this:
```julia
function product_with_err(P::ProjMPO,
                          v::ITensor)::ITensor
  Hv = v
  if isnothing(lproj(P))
    if !isnothing(rproj(P))
      Hv *= rproj(P)
    end
    for j in P.rpos-1:-1:P.lpos+1
      Hv *= P.H[j]
    end
  else #if lproj exists
    Hv *= lproj(P)
    for j in P.lpos+1:P.rpos-1
      Hv *= P.H[j]
    end
    if !isnothing(rproj(P))
      Hv *= rproj(P)
    end
  end
  order(Hv)==order(v) || throw("Order of the ProjMPO-ITensor product P*v is not equal to the order of v,"*
                               "this is probably due to an index mismatch. Try rebuilding the `ProjMPO`.")
  return noprime(Hv)
end
```

It seems to not hurt much:
```julia
using BenchmarkTools
sites = siteinds("S=1/2",10)
ampo = AutoMPO()
for j=1:length(sites)-1
  add!(ampo,1,"Sz",j,"Sz",j+1)
  add!(ampo,1,"Sx",j)
end
add!(ampo,1,"Sx",length(sites))
H= MPO(ampo,sites)

psi = productMPS(sites, fill("↑",10))
P = ProjMPO(H)
position!(P,psi,3)
wf = psi[3]*psi[4]

@btime product(P,wf) # => 22.344 μs (356 allocations: 33.38 KiB)
@btime product_with_err(P,wf) # => 22.593 μs (356 allocations: 33.38 KiB)
```

I could make a PR if this is acceptable (error message itself could be improved).

**Comment by emstoudenmire** [comment]
Hi Ori, this looks like a great way to do the check to me. Yes, computing and comparing the order of two ITensors should be extremely fast, so I’m not surprised the times are so similar. (And definitely very sub-leading compared to the other operations that ProjMPO does.) So if you could please make a PR with this change we’d appreciate it!

**Comment by mtfishman** [comment]
Closed by #390.
